Refactor `BiquadFilterNode`

b-ma opened the pull request:
Hey,

Quite a big refactoring here:
- simplified all the code
- implemented tail-time (#34) and actively processing (#147, TBC)
- added tests against frequency responses gathered from Chrome and Firefox
- support for multi-channel
- `get_frequency_response` does not rely anymore on communication between control and render threads

Performance wise: 
- There is finally no huge visible difference in the benchmark (cf. _Substractive Synth_), at the exception that it is now stereo by default while, as far as I can tell, it was only mono before, so performances of the processing itself are better. This could be improved back with #117. 
- Also all the internal computations are done using `f64`, maybe this could be moved to `f32` but then the tests against Chrome / Firefox (which internally use f46) frequency responses get wrong at some places due to floating point errors. This is something we have to decide I guess, maybe now that we are sure the coefficients and frequency response are good, we could switch back to f32 and compare against our own values to prevent regression, but maybe that's not a good idea, I don't really know. Any idea on that?

Cheers,

orottier commented:
Very nice, the previous implementation could definitely use some more concise and well-organized methods. I need some more time for the reviews, hopefully tomorrow!

orottier commented on examples/biquad.rs:
```diff
@@ -46,11 +113,9 @@ fn main() {
+    biquad.get_frequency_response(&frequency_hz, &mut mag_response, &mut phase_response);
     println!("=================================");
     println!("Biquad filter frequency response:");
     println!("=================================");
```
```
            let renderer = BiquadFilterRenderer {
                gain: g_proc,
                detune: d_proc,
                frequency: f_proc,
                q: q_proc,
                type_: type_.clone(),
                x1: Vec::with_capacity(MAX_CHANNELS),
                x2: Vec::with_capacity(MAX_CHANNELS),
                y1: Vec::with_capacity(MAX_CHANNELS),
                y2: Vec::with_capacity(MAX_CHANNELS),
            };
```

I think x1, x2, y1 and y2 are maybe more performant as `[f32; RENDER_QUANTUM_SIZE]`

b-ma commented on src/node/gain.rs:
```diff
@@ -106,7 +106,14 @@ impl AudioProcessor for GainRenderer {
+                    *value *= g;
+
+                    // flush subnormal to zero
+                    if value.is_subnormal() {
```
> FYI this branch breaks SIMD-optimizations.

Damned !

> I'm wondering if this check (and the one we just added to the AudioParam) is worth it. We are slowing down all architectures to save some performance on exotic ones?

I'm afraid they are not really exotic platforms... Intel x86_64 just like mine... ARMv8 feels more exotic to me :)

The question maybe is more: do we skip this question for no ? Open an issue to make sure we don't forget the possible problem and wait until we both better understand it and (hopefully) have a better solution to propose?

I wouldn't compare this with the check in the AudioParam that just happen once in a while in very specific conditions though

orottier replied:
Yes definitely, apparently I have some reading to do on this topic. I have created #194

b-ma commented:
> is more concise and fine for small scoped imports for me

Just removed the import altogether

> Clamping inside the loop saves an allocation (to_owned)

Nice catch!

> I think my version is a bit more clear and better for cache lines

Just replaced `all` with `any` because the filter has to continue the processing if any value in the history is not 0, but it's more clean yep!

> I think x1, x2, y1 and y2 are maybe more performant as [f32; RENDER_QUANTUM_SIZE]

Yup maybe, but I'm afraid it will add a lot of code to keep things clean when the channel number change, instead of just `self.x1.resize(num_channels, 0.);`. Actually I have an idea I'd like to test about the note I left on the clean handling of channel number change let maybe reconsider that at this moment?

b-ma commented on examples/biquad.rs:
```diff
@@ -8,28 +8,95 @@ fn main() {
+
+    std::thread::sleep(std::time::Duration::from_secs(5));
+
+    println!("> smoothly close low-pass filter for 10 sec");
```
hehe this old school lowpass effect always works :)
